feat: compiler reference documentation (#1489)
JetBrains/kotlin-web-site: pages/docs/reference/compiler-reference.md

* Enable the [progressive mode](whatsnew13.html#progressive-mode) for the compiler.
    
    In the progressive mode, deprecations and bug fixes for unstable code take effect immediately,
    instead of going through a graceful migration cycle.
    Code written in the progressive mode is backward compatible; however, code written in
    non-progressive mode may cause compilation errors in the progressive mode.
    
{:#argfile}

### `@<argfile>`
* Read the compiler options from the given file. A file passed in this option can contain compiler options with values 
and paths to source files. For example:

    <div class="sample" markdown="1" mode="shell" theme="idea">
    
    ```
    -d hello.jar -include-runtime
    hello.kt
    ```
    
    </div>

    You can also separate compiler options from source file names and pass multiple argument files.

    <div class="sample" markdown="1" mode="shell" theme="idea">
    
    ```bash
    $ kotlinc @compiler.options @classes
    ```
    
    </div>
    
    If the files reside in locations different from the current directory, use relative paths. 
    
    <div class="sample" markdown="1" mode="shell" theme="idea">
    
    ```bash

>     $ kotlinc @options\compiler.options hello.kt

Reviewer: `\` is an escape character in bash

```suggestion
    $ kotlinc @options/compiler.options hello.kt
```

    ```
    
    </div>

    
## JVM compiler options

Kotlin compiler for JVM compiles Kotlin source files into Java class files. 
The command-line tools for Kotlin to JVM compilation are `kotlinc` and `kotlinc-jvm`.
You can also use them for executing Kotlin script files.

In addition to [common options](#common-options), Kotlin/JVM compiler has the options listed below.

### `-classpath <path>` (`-cp <path>`)

> * Search for class files in the specified paths. Separate elements of the classpath with semicolons (**;**).

Reviewer: ```suggestion
* Search for class files in the specified paths. Separate elements of the classpath with system path separators (**;** on Windows, **:** on macOS/Linux).
```

The classpath can contain file and directory paths, ZIP, or JAR files.

### `-d <path>`
* Place the generated class files into the specified location. The location can be a directory, a ZIP, or a JAR file. 

### `-include-runtime`
* Include the Kotlin runtime into the resulting JAR file. Makes the resulting archive runnable on any Java-enabled 
environment.

### `-jdk-home <path>`
* Use a custom JDK home directory to include into classpath if it differs from the default `JAVA_HOME`.

### `-jvm-target <version>`
* Specify the target version of the generated JVM bytecode. Possible values are `1.6`, `1.8`, `9`, `10`, `11`, and `12`.
 The default value is `1.6`.

### `-java-parameters`
* Generate metadata for Java 1.8 reflection on method parameters.

### `-module-name <name>`
* Set the custom name for the generated `.kotlin_module` file.
  
### `-no-jdk`
* Exclude the Java runtime from the classpath.

### `-no-reflect`
* Exclude `kotlin-reflect.jar` from the classpath.
  
### `-no-stdlib`
* Exclude both `kotlin-stdlib.jar` and `kotlin-reflect.jar` from the classpath. 
  

> ### `-script <file>`

Reviewer: In fact, the `-script` argument is boolean, and it makes the compiler run the first script from the set of compiled files. (E.g. `kotlinc 1.kt 2.kts 3.kt -script` is valid, and will compile all files and run `2.kts`.)

Also note that since 1.3.70, this will be a common argument (https://github.com/JetBrains/kotlin/commit/f59e393e37d1d745261e6a2db9921b3fe5ccf7fc)

## Kotlin/JS compiler options

Kotlin compiler for JS compiles Kotlin source files into JavaScript code. 
The command-line tool for Kotlin to JS compilation is `kotlinc-js`.

In addition to [common options](#common-options), Kotlin/JS compiler has the options listed below.

### `-libraries <path>`

- Paths to Kotlin libraries with `.meta.js` and `.kjsm` files, separated by the system path separator.

### `-main {call|noCall}`

- Define whether the `main` function should be called upon execution.

### `-meta-info`

- Generate `.meta.js` and `.kjsm` files with metadata. Use this option when creating a JS library.

### `-module-kind {plain|amd|commonjs|umd}`

- Kind of the JS module generated by the compiler:
    - `plain` - a plain JS module;
    - `commonjs` - a [CommonJS](http://www.commonjs.org/) module;
    - `amd` - an [Asynchronous Module Definition](https://en.wikipedia.org/wiki/Asynchronous_module_definition) module;
    - `umd` - a [Universal Module Definition](https://github.com/umdjs/umd) module.
    
    To learn more about the module kinds and dictinctions between them,

>     read [this](https://www.davidbcalhoun.com/2014/what-is-amd-commonjs-and-umd/) article.

Reviewer: ```suggestion
    see [this](https://www.davidbcalhoun.com/2014/what-is-amd-commonjs-and-umd/) article.
```

    
### `-no-stdlib`

> - Don't use the bundled Kotlin standard library.

Reviewer: Sounds a bit ambiguous to me and too different from JVM's `-no-stdlib` option. Maybe something like this?
```
Exclude `kotlin-stdlib-js.jar` from the compilation dependencies.
```
(Maybe we should also change the description in `kotlinc-js -help`...)

Author: Yes, I'll suggest some changes to option descriptions in the Kotlin repo as well.

Author: As discussed with Zalim, this option works quite different in JS. Since the JS compiler implicitly adds stdlib as a compilation dependency, the most accurate description for the option is "Don't add" or "Don't use". So I rephrase this closer to the original description.

### `-output <path>`

> - Place the output file into the specified location.

Reviewer: Please also explain that `<path>` is the path to the resulting .js file, and not a directory for example. (There's an error in the latter case.)

In addition to [common options](#common-options), Kotlin/Native compiler has the options listed below.

### `-enable-assertions`

- Enable runtime assertions in the generated code.
    
### `-g`

- Enable emitting debug information.
    
### `-generate-test-runner` (`-tr`)

- Produce a runner for unit tests.
    
### `-generate-worker-test-runner` (`-trw`)

- Produce a worker runner for unit tests.
    
### `-generate-no-exit-test-runner` (`-trn`)

- Produce a runner for unit tests not forcing exit.
    
### `-include-binary <path>` (`-ib <path>`)

- Pack external binary within the generated klib file.
    
### `-library <path>` (`-l <path>`)

- Link with the library.

### `-library-version <version>` (`-lv`)

- Set library version.
    
### `-list-targets`

- List available hardware targets.

### `-manifest <path>`

- Provide a manifest addend file.
    
### `-memory-model {strict|relaxed}`

- Memory model to use. Two memory models are supported:
    - _strict_
    - _relaxed_

### `-module-name <name>`

- Specify a name for the compilation module.

### `-native-library <path>`(`-nl <path>`)

- Include the native bitcode library.

### `-no-default-libs`

> - Don't link the libraries from dist/klib automatically.

Reviewer 2: Description of this flag is obsolete and should be update at the compiler side. Currently it disables linking a user code with default platform libraries distributed with the compiler (More about platform libraries: https://kotlinlang.org/docs/reference/native/platform_libs.html).
